core/src/main/java/io/questdb/cairo/wal/seq/TableSequencerImpl.java

```diff
@@ -0,0 +1,280 @@
+import io.questdb.griffin.SqlException;
+import io.questdb.log.Log;
+import io.questdb.log.LogFactory;
+import io.questdb.std.FilesFacade;
+import io.questdb.std.Misc;
+import io.questdb.std.SimpleReadWriteLock;
+import io.questdb.std.str.Path;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.concurrent.locks.ReadWriteLock;
+
+import static io.questdb.cairo.wal.WalUtils.WAL_INDEX_FILE_NAME;
+
+public class TableSequencerImpl implements TableSequencer {
+    private static final Log LOG = LogFactory.getLog(TableSequencerImpl.class);
+    private final ReadWriteLock schemaLock = new SimpleReadWriteLock();
+    private final CairoEngine engine;
+    private final String tableName;
+    private final int rootLen;
+    private final SequencerMetadata metadata;
+    private final TableTransactionLog tableTransactionLog;
+    private final IDGenerator walIdGenerator;
+    private final Path path;
+    private final BinaryAlterFormatter alterCommandWalFormatter = new BinaryAlterFormatter();
+    private final SequencerMetadataUpdater sequencerMetadataUpdater;
+    private final FilesFacade ff;
+    private final int mkDirMode;
+    private volatile boolean open = false;
+    private boolean distressed;
+
+    TableSequencerImpl(CairoEngine engine, String tableName) {
+        this.engine = engine;
+        this.tableName = tableName;
+
+        final CairoConfiguration configuration = engine.getConfiguration();
+        final FilesFacade ff = configuration.getFilesFacade();
+        try {
+            path = new Path();
+            path.of(configuration.getRoot()).concat(tableName).concat(WalUtils.SEQ_DIR);
+            rootLen = path.length();
+            this.ff = ff;
+            this.mkDirMode = configuration.getMkDirMode();
+
+            metadata = new SequencerMetadata(ff);
+            sequencerMetadataUpdater = new SequencerMetadataUpdater(metadata, tableName);
+            walIdGenerator = new IDGenerator(configuration, WAL_INDEX_FILE_NAME);
+            tableTransactionLog = new TableTransactionLog(ff);
+        } catch (Throwable th) {
+            LOG.critical().$("could not create sequencer [name=").$(tableName)
+                    .$(", error=").$(th.getMessage())
+                    .I$();
+            doClose();
+            throw th;
+        }
+    }
+
+    @Override
+    public void close() {
+        schemaLock.writeLock().lock();
+        try {
+            doClose();
+        } finally {
+            schemaLock.writeLock().unlock();
+        }
+    }
+
+    @Override
+    public void copyMetadataTo(@NotNull SequencerMetadata copyTo) {
+        schemaLock.readLock().lock();
+        try {
+            copyTo.copyFrom(metadata);
+        } finally {
+            schemaLock.readLock().unlock();
+        }
+    }
+
+    @Override
+    public TableMetadataChangeLog getMetadataChangeLogCursor(long structureVersionLo) {
+        checkDistressed();
+        if (metadata.getStructureVersion() == structureVersionLo) {
```

Comment:
Also racy.

Reply:
Marking `SequencerMetadata#structureVersion` as `volatile` would help only if we're ok with potential races here, i.e. `getMetadataChangeLogCursor()` may return a no-op cursor while the latest version is already above `structureVersionLo`.

Reply:
This also is an optimisation, if left side is out of date empty cursor will be returned. 

I changed metadata cursor implementation that it returns back empty cursor in case it cannot find the records in the file in case `metadata.getStructureVersion()` returns higher version than is visible in the files.

Reply:
In any case, `structureVersion` should be `volatile`.

Reply:
why, if we can handle it's out of date and it changes infrequently?

Reply:
Because there is no guarantee that this check will ever work, so we can remove it completely with the same effect.

Reading an uncontested `volatile` field is much cheaper than you'd expect: https://brooker.co.za/blog/2012/09/10/volatile.html

Reply:
We can remove completely, it's fine. If fact it's removed now. The code which use cursor go in and out of locks and other volatile variables, memory barriers happen frequently.

Reply:
> If fact it's removed now.

Thanks!